File: perf/benchmarks/apps/spectral_norm/spectral_norm.cpp
// Spectral norm — Computer Language Benchmarks Game numerical kernel (C++).
//
// Algorithmically identical to the Python and Cangjie versions: 10 power
// iterations of v <- A^T A v on the same A(i,j) infinite-matrix function,
// same array layout, same final checksum sqrt(<u,v>/<v,v>) * 1e9.
#include <chrono>
#include <cmath>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <vector>

static inline double A(int i, int j) {
    return 1.0 / static_cast<double>(((i + j) * (i + j + 1) >> 1) + i + 1);
}

static void Av(const std::vector<double>& v, std::vector<double>& out, int n) {
    for (int i = 0; i < n; ++i) {
        double s = 0.0;
        for (int j = 0; j < n; ++j) s += A(i, j) * v[static_cast<size_t>(j)];
        out[static_cast<size_t>(i)] = s;
    }
}

static void Atv(const std::vector<double>& v, std::vector<double>& out, int n) {
    for (int i = 0; i < n; ++i) {
        double s = 0.0;
        for (int j = 0; j < n; ++j) s += A(j, i) * v[static_cast<size_t>(j)];
        out[static_cast<size_t>(i)] = s;
    }
}

static void AtAv(const std::vector<double>& v, std::vector<double>& out,
                 std::vector<double>& tmp, int n) {
    Av(v, tmp, n);
    Atv(tmp, out, n);
}

int main(int argc, char** argv) {
    int n = (argc > 1) ? std::atoi(argv[1]) : 1500;

    std::vector<double> u(static_cast<size_t>(n), 1.0);
    std::vector<double> v(static_cast<size_t>(n), 0.0);
    std::vector<double> tmp(static_cast<size_t>(n), 0.0);

    auto t0 = std::chrono::steady_clock::now();
    for (int i = 0; i < 10; ++i) {
        AtAv(u, v, tmp, n);
        AtAv(v, u, tmp, n);
    }
    double vbv = 0.0, vv = 0.0;
    for (int i = 0; i < n; ++i) {
        vbv += u[static_cast<size_t>(i)] * v[static_cast<size_t>(i)];
        vv  += v[static_cast<size_t>(i)] * v[static_cast<size_t>(i)];
    }
    double sn = std::sqrt(vbv / vv);
    auto t1 = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();

    int64_t cs = static_cast<int64_t>(std::llround(sn * 1e9));
    std::printf("CHECKSUM:%lld\n", static_cast<long long>(cs));
    std::printf("ELAPSED_MS:%.6f\n", ms);
    return 0;
}

File: perf/benchmarks/apps/spectral_norm/spectral_norm.py
# Spectral norm — Computer Language Benchmarks Game numerical kernel.
#
# Approximates the largest eigenvalue of an infinite matrix A with
# A[i][j] = 1 / ((i+j)*(i+j+1)/2 + i + 1) by 10 power-iteration steps of
# v <- A^T * A * v. Highly symmetric across all three implementations:
# same matrix function, same number of iterations, same array layout,
# same checksum (sqrt(<u,v>/<v,v>)).
#
# Argument: N (default 1500).

import math
import sys
import time


def A(i: int, j: int) -> float:
    return 1.0 / (((i + j) * (i + j + 1) >> 1) + i + 1)


def Av(v: list, out: list, n: int) -> None:
    for i in range(n):
        s = 0.0
        for j in range(n):
            s += A(i, j) * v[j]
        out[i] = s


def Atv(v: list, out: list, n: int) -> None:
    for i in range(n):
        s = 0.0
        for j in range(n):
            s += A(j, i) * v[j]
        out[i] = s


def AtAv(v: list, out: list, tmp: list, n: int) -> None:
    Av(v, tmp, n)
    Atv(tmp, out, n)


def main() -> None:
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 1500

    u = [1.0] * n
    v = [0.0] * n
    tmp = [0.0] * n

    t0 = time.perf_counter_ns()
    for _ in range(10):
        AtAv(u, v, tmp, n)
        AtAv(v, u, tmp, n)
    vbv = 0.0
    vv = 0.0
    for i in range(n):
        vbv += u[i] * v[i]
        vv += v[i] * v[i]
    sn = math.sqrt(vbv / vv)
    elapsed_ns = time.perf_counter_ns() - t0

    cs = round(sn * 1e9)
    print(f"CHECKSUM:{cs}")
    print(f"ELAPSED_MS:{elapsed_ns / 1_000_000:.6f}")


if __name__ == "__main__":
    main()
